File: AzanAlarm-RN/src/services/QiblaService.ts
// Mecca coordinates
const MECCA_LAT = 21.4225;
const MECCA_LNG = 39.8262;

/**
 * Calculate Qibla direction (bearing to Mecca) from given coordinates
 * Returns bearing in degrees (0-360)
 */
export const calculateQiblaDirection = (latitude: number, longitude: number): number => {
    const lat1 = toRadians(latitude);
    const lng1 = toRadians(longitude);
    const lat2 = toRadians(MECCA_LAT);
    const lng2 = toRadians(MECCA_LNG);

    const dLng = lng2 - lng1;

    const y = Math.sin(dLng) * Math.cos(lat2);
    const x = Math.cos(lat1) * Math.sin(lat2) - Math.sin(lat1) * Math.cos(lat2) * Math.cos(dLng);

    let bearing = toDegrees(Math.atan2(y, x));
    bearing = (bearing + 360) % 360; // Normalize to 0-360

    return bearing;
};

/**
 * Calculate distance to Mecca in kilometers
 */
export const calculateDistanceToMecca = (latitude: number, longitude: number): number => {
    const R = 6371; // Earth's radius in km

    const lat1 = toRadians(latitude);
    const lng1 = toRadians(longitude);
    const lat2 = toRadians(MECCA_LAT);
    const lng2 = toRadians(MECCA_LNG);

    const dLat = lat2 - lat1;
    const dLng = lng2 - lng1;

    const a =
        Math.sin(dLat / 2) * Math.sin(dLat / 2) +
        Math.cos(lat1) * Math.cos(lat2) * Math.sin(dLng / 2) * Math.sin(dLng / 2);

    const c = 2 * Math.atan2(Math.sqrt(a), Math.sqrt(1 - a));

    return R * c;
};

const toRadians = (degrees: number): number => {
    return degrees * (Math.PI / 180);
};

const toDegrees = (radians: number): number => {
    return radians * (180 / Math.PI);
};

File: internal/services/qibla_service.go
// Package services contains business logic for the AzanAlarm application.
package services

import (
	"math"
)

// QiblaService calculates the Qibla direction.
type QiblaService struct{}

// Makkah coordinates (Kaaba)
const (
	MakkahLatitude  = 21.4225
	MakkahLongitude = 39.8262
)

// NewQiblaService creates a new QiblaService instance.
func NewQiblaService() *QiblaService {
	return &QiblaService{}
}

// GetQiblaDirection calculates the Qibla direction from a given location.
// Returns the bearing in degrees from North (0-360).
func (qs *QiblaService) GetQiblaDirection(latitude, longitude float64) float64 {
	// Convert to radians
	lat1 := qs.degreesToRadians(latitude)
	lon1 := qs.degreesToRadians(longitude)
	lat2 := qs.degreesToRadians(MakkahLatitude)
	lon2 := qs.degreesToRadians(MakkahLongitude)

	// Calculate bearing using great circle formula
	dLon := lon2 - lon1

	x := math.Sin(dLon) * math.Cos(lat2)
	y := math.Cos(lat1)*math.Sin(lat2) - math.Sin(lat1)*math.Cos(lat2)*math.Cos(dLon)

	bearing := math.Atan2(x, y)
	bearingDegrees := qs.radiansToDegrees(bearing)

	// Normalize to 0-360
	if bearingDegrees < 0 {
		bearingDegrees += 360
	}

	return bearingDegrees
}

// GetDistanceToMakkah calculates the distance to Makkah in kilometers.
func (qs *QiblaService) GetDistanceToMakkah(latitude, longitude float64) float64 {
	// Earth radius in kilometers
	const R = 6371.0

	lat1 := qs.degreesToRadians(latitude)
	lon1 := qs.degreesToRadians(longitude)
	lat2 := qs.degreesToRadians(MakkahLatitude)
	lon2 := qs.degreesToRadians(MakkahLongitude)

	dLat := lat2 - lat1
	dLon := lon2 - lon1

	// Haversine formula
	a := math.Sin(dLat/2)*math.Sin(dLat/2) +
		math.Cos(lat1)*math.Cos(lat2)*math.Sin(dLon/2)*math.Sin(dLon/2)
	c := 2 * math.Atan2(math.Sqrt(a), math.Sqrt(1-a))

	return R * c
}

// degreesToRadians converts degrees to radians.
func (qs *QiblaService) degreesToRadians(degrees float64) float64 {
	return degrees * math.Pi / 180.0
}

// radiansToDegrees converts radians to degrees.
func (qs *QiblaService) radiansToDegrees(radians float64) float64 {
	return radians * 180.0 / math.Pi
}
